### PR TITLE
Allow players to drop unknown items and add a setting to remove them on drop.

### PR DESCRIPTION
**Pull request by red-001**
See:  #3389

### CODE REVIEW

**Comment by ObaniGemini** [comment]
:-1: sometimes unknown items appear after a bad server update. If it removes important nodes for the player, I'm against it. Or make it be an option in minetest.conf

**Comment by MoNTE48** [comment]
+1!

**Comment by nerzhul** [comment]
:-1: if no option to configure this behaviour

**Comment by red-001** [comment]
@ObaniGemini  @nerzhul  Is it better now?

File: src/script/lua_api/l_env.cpp [hunk 1]
```diff
@@ -35,9 +35,10 @@ with this program; if not, write to the Free Software Foundation, Inc.,
 #include "treegen.h"
 #include "emerge.h"
 #include "pathfinder.h"
-
+#include "settings.h"
 ///////////////////////////////////////////////////////////////////////////////
 
+extern Settings *g_settings;
```
Comment (nerzhul) [inline src/script/lua_api/l_env.cpp]: Why there is this here ? it's not needed

File: src/script/lua_api/l_env.cpp [hunk 2]
```diff
@@ -420,8 +421,12 @@ int ModApiEnvMod::l_add_item(lua_State *L)
 	//v3f pos = checkFloatPos(L, 1);
 	// item
 	ItemStack item = read_item(L, 2,getServer(L));
-	if(item.empty() || !item.isKnown(getServer(L)->idef()))
+	if(item.empty())
```
Comment (nerzhul) [inline src/script/lua_api/l_env.cpp]: space

**Comment by nerzhul** [comment]
Please fix the two last isssues mentionned and then :+1:

File: src/script/lua_api/l_env.cpp (hunk 1 above)
Comment (red-001) [inline src/script/lua_api/l_env.cpp]: The code doesn't work correctly without it.

**Comment by ObaniGemini** [comment]
Is everything ok for this ?

File: src/script/lua_api/l_env.cpp (hunk 1 above)
Comment (red-001) [inline src/script/lua_api/l_env.cpp]: I tested it again and I worked without this.

**Comment by paramat** [comment]
:-1: Unnecessary setting, i will support enabling the dropping of unknown nodes but they should always disappear on drop, as the original behaviour seems to be. Is there a need to drop an unknown and keep it in the world?

**Comment by 0-afflatus** [comment]
> Is there a need to drop an unknown and keep it in the world?

I can't think of one.

**Comment by ObaniGemini** [comment]
@paramat : this is a stupid question. If the server admin removed a texture or a node by error, it will be as an unknown block. If he know what he does it's ok but if he wants a maximum of security, he can disable the setting, to re-add missing nodes/textures that he forgot to add.
This is way more secure, and clients still can erase the unknown blocks by themselves.

**Comment by paramat** [comment]
I'm not comvinced, let's choose the best behaviour, whatever that is, and not have a setting.

**Comment by ObaniGemini** [comment]
The best behaviour would be to keep it as a setting. Most cases are a texture or a node forgotten at the server reboot.
But if a node is erased and the host doesn't do any aliases, it's good to have this setting.

Every server hoster doing his work should know that.

**Comment by red-001** [comment]
@ObaniGemini A forgotten texture would just make minetest generate a placeholder.

**Comment by ObaniGemini** [comment]
It's not just about unknown textures

**Comment by red-001** [comment]
Closing as it's unlikely that this will get merged soon

**Pull request closed without merging**